src/Microsoft.Net.Http.Headers/NameValueHeaderValue.cs

```diff
@@ -142,6 +147,20 @@ public override bool Equals(object obj)
             }
         }
 
+        public StringSegment GetDecodedValue()
+        {
+            return HeaderUtilities.RemoveQuotes(_value);
+        }
+
+        public void SetAndEncodeValue(StringSegment value)
+        {
+            if (_quotedValue)
+            {
+                value = $"\"{value}\"";
```

Comment:
validate as token
if not token, validate as quoted string
if not quoted string, add quotes and escape reserved chars.
validate as quoted string, throw if still not valid (not everything can be escaped).